authentication/azure/storage.go

```diff
@@ -0,0 +1,61 @@
+package azure
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/Azure/azure-storage-blob-go/azblob"
+	"github.com/Azure/go-autorest/autorest/azure"
+
+	"github.com/dapr/kit/logger"
+)
+
+const (
+	storageAccountKeyKey = "accountKey"
+)
+
+// GetAzureStorageCredentials returns a azblob.Credential object that can be used to authenticate an Azure Blob Storage SDK pipeline.
+// First it tries to authenticate using shared key credentials (using an account key) if present. It falls back to attempting to use Azure AD (via a service principal or MSI).
+func GetAzureStorageCredentials(log logger.Logger, accountName string, metadata map[string]string) (azblob.Credential, *azure.Environment, error) {
+	settings, err := NewEnvironmentSettings("storage", metadata)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Try using shared key credentials first
+	accountKey, ok := metadata[storageAccountKeyKey]
+	if ok && accountKey != "" {
+		credential, newSharedKeyErr := azblob.NewSharedKeyCredential(accountName, accountKey)
+		if err != nil {
+			return nil, nil, fmt.Errorf("invalid credentials with error: %s", newSharedKeyErr.Error())
+		}
+
+		return credential, settings.AzureEnvironment, nil
+	}
+
+	// Fallback to using Azure AD
+	spt, err := settings.GetServicePrincipalToken()
+	if err != nil {
+		return nil, nil, err
+	}
+	var tokenRefresher azblob.TokenRefresher = func(credential azblob.TokenCredential) time.Duration {
+		log.Debug("Refreshing Azure Storage auth token")
+		fmt.Println("Refreshing Azure Storage auth token")
```

Comment:
I just noticed there are 2 log statements here. Perhaps only one of the two should be kept (or neither). Same on lines 53-54. Not sure what the standard for logging in Dapr is, maybe you could decide how to fix this please?

Reply:
We shouldn't use fmt.Println

Reply:
I already removed it before I saw your comment @yaron2 :)